pandas/tests/indexes/test_base.py

```diff
@@ -751,37 +759,66 @@ def test_union(self):
         first = Index(list('ab'))
         second = Index(list('ab'), name='B')
         union = first.union(second)
-        expected = Index(list('ab'), name='B')
+        expected = Index(list('ab'), name=None)
+        tm.assert_index_equal(union, expected)
+
+        # GH 9943 9862
+        first = Index(list('abc'))
+        second = Index(list('ab'), name='B')
+        union = first.union(second)
+        expected = Index(list('abc'), name=None)
         tm.assert_index_equal(union, expected)
 
         first = Index([])
         second = Index(list('ab'), name='B')
         union = first.union(second)
-        expected = Index(list('ab'), name='B')
+        expected = Index(list('ab'), name=None)
         tm.assert_index_equal(union, expected)
 
         first = Index(list('ab'))
         second = Index([], name='B')
         union = first.union(second)
-        expected = Index(list('ab'), name='B')
+        expected = Index(list('ab'), name=None)
         tm.assert_index_equal(union, expected)
 
         first = Index(list('ab'), name='A')
         second = Index(list('ab'))
         union = first.union(second)
-        expected = Index(list('ab'), name='A')
+        expected = Index(list('ab'), name=None)
+        tm.assert_index_equal(union, expected)
+
+        # GH 9943 9862
+        first = Index(list('ab'), name='A')
+        second = Index(list('abc'))
+        union = first.union(second)
+        expected = Index(list('abc'), name=None)
         tm.assert_index_equal(union, expected)
 
         first = Index(list('ab'), name='A')
         second = Index([])
         union = first.union(second)
-        expected = Index(list('ab'), name='A')
+        expected = Index(list('ab'), name=None)
         tm.assert_index_equal(union, expected)
 
         first = Index([], name='A')
         second = Index(list('ab'))
         union = first.union(second)
-        expected = Index(list('ab'), name='A')
+        expected = Index(list('ab'), name=None)
```

Comment:
same

Reply:
@jreback Not sure what you mean by "make new tests".  The issue with the existing tests is that they were incorrect.  So I had to fix them.  For example, 
```
first = Index([], name='A')
second = Index(list('ab'))
union = first.union(second)
expected = Index(list('ab'), name='A')
```
is incorrect, as if one index has a name, and the other has no name, the result should be `None`, based on your comment https://github.com/pandas-dev/pandas/issues/9943#issuecomment-368163701

Reply:
I meant that these tests are getting very long, so split up a bit.

We also need more general testing, *all* of these tests should be using a generically created index (all of this machinery already exists ia .create_index())

Reply:
I've split the `test_union` test, and now have the corner cases tested for all indexes